File: backend/app/schemas/document.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.models.document import DocumentStatus


class DocumentRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    original_filename: str
    mime_type: str
    file_extension: str
    file_size: int
    status: str
    processing_error: str | None
    indexed_at: datetime | None
    created_at: datetime
    updated_at: datetime


class DocumentDetailRead(DocumentRead):
    chunk_count: int = 0


class DocumentListResponse(BaseModel):
    documents: list[DocumentRead]
    total: int


class DocumentUploadResponse(BaseModel):
    document: DocumentRead
    message: str


class DocumentDeleteResponse(BaseModel):
    id: uuid.UUID
    message: str


class DocumentSearchRequest(BaseModel):
    query: str = Field(
        min_length=1,
        max_length=2000,
    )
    top_k: int | None = Field(
        default=None,
        ge=1,
        le=50,
    )

    @field_validator("query")
    @classmethod
    def normalize_query(cls, value: str) -> str:
        normalized = " ".join(value.split())

        if not normalized:
            raise ValueError("Search query cannot be empty")

        return normalized


class DocumentSearchResult(BaseModel):
    chunk_id: uuid.UUID
    document_id: uuid.UUID
    filename: str
    chunk_index: int
    content: str
    page_number: int | None
    source: str | None
    similarity: float


class DocumentSearchResponse(BaseModel):
    query: str
    results: list[DocumentSearchResult]
    total: int


class DocumentStatusResponse(BaseModel):
    id: uuid.UUID
    status: str
    processing_error: str | None
    indexed_at: datetime | None

    @field_validator("status")
    @classmethod
    def validate_status(cls, value: str) -> str:
        allowed = {
            DocumentStatus.PROCESSING.value,
            DocumentStatus.INDEXED.value,
            DocumentStatus.FAILED.value,
        }

        if value not in allowed:
            raise ValueError("Invalid document status")

        return value

File: frontend/src/types/document.ts
export type DocumentStatus =
  | 'processing'
  | 'indexed'
  | 'failed'


export interface DocumentRecord {
  id: string
  original_filename: string
  mime_type: string
  file_extension: string
  file_size: number
  status: DocumentStatus
  processing_error: string | null
  indexed_at: string | null
  created_at: string
  updated_at: string
}


export interface DocumentDetail
  extends DocumentRecord {
  chunk_count: number
}


export interface DocumentListResponse {
  documents: DocumentRecord[]
  total: number
}


export interface DocumentUploadResponse {
  document: DocumentRecord
  message: string
}


export interface DocumentDeleteResponse {
  id: string
  message: string
}


export interface DocumentStatusResponse {
  id: string
  status: DocumentStatus
  processing_error: string | null
  indexed_at: string | null
}


export interface DocumentSearchInput {
  query: string

  /**
   * Maximum final number of parent-level search results.
   *
   * Backend currently accepts 1..20.
   */
  top_k?: number
}


export interface DocumentSearchResult {
  /**
   * ID of the highest-ranked child chunk that caused
   * this parent context to be selected.
   */
  chunk_id: string

  document_id: string

  filename: string

  chunk_index: number

  /**
   * Parent-level content returned after child retrieval,
   * CrossEncoder reranking and parent expansion.
   */
  content: string

  page_number: number | null

  source: string | null

  /**
   * Initial pgvector cosine similarity.
   */
  similarity: number

  /**
   * Final CrossEncoder relevance score.
   *
   * null is possible when reranking is disabled.
   */
  rerank_score: number | null
}


export interface DocumentSearchResponse {
  query: string

  results: DocumentSearchResult[]

  total: number
}